RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+
+
+
+* **KSR_0003**: the watchdog subsystem shall ensure the opened WD device to be the one specified as input argument
+* **KSR_0004**: The watchdog subsystem shall ensure the WD timeout to be set according to the IOCTL input parameter
+* **KSR_0005**: The operating system shall ensure the WD timeout to be not wrongly re-set to a different timeout value
+* **KSR_0012**: Writing the Watchdog shall reset it with a timeout equal to the one specified in the IOCTL. No other timeout value shall be used / reconfigured
+
+**Assumption:** it is assumed that any boot time operation has been successfully completed; we could consider the design of a RV Monitor for guarding against boot time misconfigurations as a follow-on activity
+
+
+### Static and Dynamic Architecture
+
+![alt_text](./i6300esb_watchdog_timeout.png "image_tooltip")
+ \
+As a first step we consider only the architectural elements that are functionally supporting the implementation of the safety requirements mentioned above; interfering architectural elements will be added later in the picture. \
+ \
+In the use case under analysis, to start with, we can identify 3 main elements interacting with each other:
+
+
+
+* The Safety Application
+* The Linux Kernel
+* The HW SoC that also includes a programmable i6300esb watchdog.
+
+ \
+Accordingly we have the following events that characterize the expected interactions in absence of failures.
+
+
+
+1. The Safety Application invokes the open syscall passing the input path from the filesystem associated with the watchdog to be opened.
+2. Based on the path passed from open, the Kernel:
+    1. Allocates a unique file descriptor (fd) associated with the such watchdog device;
+    2. Unlocks the watchdog registers for write access;
+    3. Reloads the external watchdog (stops any previous counting) 
+    4. Starts the watchdog using the default timeout programmed at boot time;
+3. In case of success the Kernel returns the allocated unique fd to the Safety Application
+4. The safety application uses the returned fd to set a new watchdog timeout by invoking  `ioctl(fd, WDIOC_SETTIMEOUT, &timeout`); where:
+    5. fd is the watchdog file descriptor returned by open
+    6.  `WDIOC_SETTIMEOUT `is the ioctl command that corresponds to the operation of setting a watchdog timeout
+    7.  `&timeout `is the user space address of the variable containing the timeout value to be written to the watchdog
+5. According to the fd passed by the Safety Application the Kernel:
+    8. Shift Left by 9 bits the timeout value;
+    9. Unlocks the watchdog registers for write access;
+    10. Write the shifted value to `ESB_TIMER1_REG;`
+    11. Unlocks the watchdog registers for write access;
+    12. Write the same shifted value to `ESB_TIMER2_REG;`
+    13. Unlocks the watchdog registers for write access;
+    14. Reloads the external watchdog (stops any previous counting)
+6. If successful the Kernels returns 0 to the Safety Application, else -1 is returned.
```

Comment:
So, basically, till here, this might work for re-configuring the WD, but strictly speaking, it is not part of the safe initialisation and it could be omitted, I think, because per-se it is not sufficient to claim that the WD was initialised safel.

Reply:
I don't understand....here the Kernel programs the WTD with the timeout passed by the safety app and restart the WTD timer. If it doesn't program the WTD with the right value OR it does not kick off the WTD timer our safety mechanism (the WTD) will not work as expected and the safety app will continue un-monitored....

Reply:
Exactly!
My point is that the WD needs to have been initialised by whatever started the safe app, not the safe app itself.
The safe app should be allowed - possiby - to reprogram the WD ONLY after it has been proven that the allegedly safe app is indeed safe, in this specific round of execution. But in a simpler scenario the safe app doesn't program the WD, which gets setup by the safe component that spawned the safe app.

Reply:
But by definition, considering how the program is structured, the safety APP does not start the actual safe operations till the ioctl returns success, so if, for any reason, the App never comes to the point of ioctl returning success, there would be no harm involved

Reply:
I am trying to describe how a more generic system should be designed.

What we have here is a corner case. It can be described as a corner case, taking advantage of its corner-casyness, of course.

But how useful is it? To a reader who probably needs to adapt this to something else?

I'd rather have the corner case to still be implemented in a generic way and take advantage of being "special" only as the last resort, to get out of a problem that could otherwise not be solved. 

This is definitely not the case.

Reply:
As we discussed also during the WG, there isn't much we can do about this point, else we should consider a completely different use case

Reply:
https://simple.wikipedia.org/wiki/Sunk_cost_fallacy

Reply:
It is not about the time spent, it is about the time it would take, now, to start on a different, more interesting, use case before being able to come to the same level of detail that would allow us to desing an RV Monitor and an associated safety analysis. As I said already, let's get this done now and then we move on

Reply:
I see two issues
1) Technical advandage
Let's assume that I follow you suggestion and, for example, the RV Monitor is already aware of the timeout, that would be programmed by some FW, instead of it being passed by ioctl.
Now in this case the safety app will
a) open the wtd  
b) regularly pet it, if everything is OK. 

Without a) we cannot get the fd to pet it, right?

Ok but now, what happens is that, when the i6300esb driver is [probed](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/i6300esb.c#L309) the hard coded ESB_HEARTBEAT_DEFAULT default timeout is set in wdd.timeout or, if the driver is dynamically loaded, such value is [set](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/i6300esb.c#L310) according to the module param 

Then as part of the init process this value is [programmed](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/i6300esb.c#L286) inside the actual watchdog, hence overwriting the value previously programmed by FW with the timer being restarted.

So technically we should rework the safety app to:
a) modprobe the WTD driver (with the same timeout as used by FW)
b) open the WTD
c) write it regularly
TBH I do not see less complexity compared to our actual use case

2) Productivity of the WG. IMO we are close to complete an example of a safety claim support (with assumptions that you think are not making this example credible) but that still describe a methodology that can be extended to other, more interesting, safety claims. So, as also agreed, as part of the WG roadmap I'd prefer to close this one and then move one to more interesting cases (and please feel free to chime in with proposals)

p.s. in our meeting we discussed about evaluating gemini to produce a doc baseline to be humanly refined. Improving the Kernel doc is beneficial to everybody. Technically I could also use the safety analysis template filling it with the info I already know, but it would not be as beneficial to the community
p.p.s I really encourage to have an offline 1:1 to agree on how to close this point OR we can discuss this in the next WG meeting

Reply:
Ok, I see I have not been sufficiently clear on various aspects of my predicament; let's see if I cna fix that.

**Complexity**
There are actualyl 2 types of complexity:

1. the complexity of the implementation
2. the complexity in allocating and meeting safety requirements

My comment was mostly about the second point. Why?

Because in the case where the app sets up the watchdog for itself, there are much stronger requirements applied to almost everything. Precisely those where we got swamped.

In the cccase as you have attempted to implement it, you have a safety critical section where the application is supposed to configure correctly the WD for itself, so that it will catch later failured.

This approach presents several weaknesses and vulnerability:

- The application can fail in setting up the WD and not detect it, then proceed using it as if it was functioning correctly. You might be tempted to argue that these are multiplle failures and they are not considered, but allow me to retort, that - being the very same component, the WD, it is highely likely that they are correlated, therefore the "multiple failure" counter argument is moot.

- You have to prove that the interaction with the kernel (at least for the initial configuration of the WD) is free from interference. Which is in se already a quite bold claim to make and would deserve a separate discussion. But I hope you will agree that this is something that needs to be considered, in the safety analysis.
Besides, all of this happens at a point of the boot phase where potentially many things have already happeend, so you cannot even maek a realistic claim that - being at the very beginning of the boot - you can be relatively confident not much has happeend yet, that could interfere.

-  The interaction with the WD relies on a correct implementation of the WD driver, so now you have also this initialisation and the associated kernel code that are at the receiving end of safety requirements.
Incidentally, this is where we got stuck.


Let's look at the other case:

1. something that can be considered safe primes the WD with an adequate time
2. events are put in motion that are expected to culminate in starting the app
3. the app is expected to do what it needs to do, to ping the WD

Et voila', the only part which receives safety requirements (from an FFI perspective) is point #1
Of course, in case there are also requirements about availability, then the points 2 and 3 must be developed adequately. But only in that case. And here we were not covering it, AFAIK.

I know it now seems like I'm hiding behind point 1, but in reality, Linux is NOT safe per se', nor it can be made easily safe, and it would be misleading to claim otherwise.

One should deal with allocation of safety requirement to Linux only where it is absolutely impossible to do otherwise. Here, clearly, it is possible to avoid it.

Ad of course there is still an effort in making the bootloader or the hypervisor safe. Or perhaps one will use a WD that can be configured through external pins. I do not know.

To recap, the whole approach is misleading, because it seems to use as excuse "we in ELISA do not offer guarrantees" to dodge the hard problems.
I'm not saying that it's intentional, but how useful it is to someone ading these docs, if the solution proposed is impractical or unapplicable?

The first step os to assess some "first principle", some ground truth.
Picking the right HW is one of them. This doesn't absolutely mean that one has to choose a specific component, but it needs to be possible to allocate some requirements on it.

For example: is that i6300esb a good choice? I'm not so sure. A good WD should allow for various options, including not being disabled and not allowing the period to be re-programmed to be longer, for example.
Why do you assume that the WD driver will be loaded by the application?
Why do you assume that it will not be modified as needed?

Regarding the productiity of the WG: I was assuming that producing correct results was more important than producing a volume of results, but perhaps I got the KPI wrong :-D 

I'm more interest in pursuing what is correct. If that is not aligned, please let me know.

In particular, if the methodology doesn't start with attempting to remove Linux from the allocation of safety constraints, and setting some basic truth, like relying on a WD that was primed by a safe component, I doubt it's particularly useful. Or desirable.
Even these sort of considerations should be considered a valid result, IMHO, because they document various approaches and can recommend which ones are mor exposed to risks.


I would suggest booking a session with the OSEP WG to discuss the overall process and method.

Reply:
I would like to close this comment tomorrow in the Safety Architecture WG; however in the meantime I'll provide some answers:
- why did we choose the i6300esb WTD? Because it was the available HW also used by the Automotive WG and by the AGL project. Such HW was used to build the telltale demo on top of AGL. Could we use a better one? Yes if we had an automotive platform Avalaible with related doc, but at the time we didn't have a better choice. If in future maybe NVidia provides HW platform and associated programming doc available we can definitively use something more meaningful
- Am I focusing on producing a lot of doc rather than meaningful doc? No. I am focusing on a viable methodology to analyse the Kernel and, in this case, with the goal of designing a Runtime Verification Monitor. IMO, working on the current assumptions, that are based on claims that we assume but we did not demonstrate, would not make any difference on the methodology we use to analyse the Kernel. As I said, then we can leverage such methodology to also cover more useful use cases (including the topics proposed in the OSEP WG)
- Should I sync up with OSEP? Yes of course. My intention is to complete a first draft of the safety analysis for the RVM design and then discuss with OSEP such methodology and refine it. If Paul wants to have a joint session before we even complete the first pass I open to it too

Reply:
>  If in future maybe NVidia provides HW platform and associated programming doc available we can definitively use something more meaningful

LOL
My point is that one should be able to define what to expect from the HW, and select the HW accordingly, rather than picking one thing and then try to make it work, because it is not guarranteed that it will. Not even after bending backward several times.

Using Linux in a safety scenario DOES impose limitations, and specifying, for example, that a certain behaviour or a certain safety qualification is expected from other components, would be a very good way to point to these limitations.

Reply:
From a use case perspective, the arch wg relies on domain WGs to define a safety concept and allocate requirements on the OS.
Now if you roll back in history you will realise that telltale was the only completed TSC from the automotive WG. What I always said in every forum is that the Arch WG will keep investigating methodologies to analyse and qualify the Kernel given a set of safety requirements and then, when a more realistic use case is available, the WG will re-align to it.

So from my perspective you are stating the obvious. What you are missing instead is that, even under assumptions of claims to be demonstrated, the methodology we are working on is still useful and scalable later to other more realistic use cases.

Reply:
eh, initially I did not arue against this because it looked like this might be made to work ... 

but it seems now clear that it will not reach a real conclusion, because of lak of documentation

sure, one can argue that the documentation can be enhanced, but even doing that, the calims will then rest on a variety of avoidable burdens

so, I wonder, why not just fixing this?

but hey, these are just my thoughts and the arguments supporting them

if you think they are wrong or inconsistent, feel free to ignore them, you don't have to convince me :)

Reply:
Igor AFAIK the Automotive WG tried to revisit the telltale use case to match [something](https://github.com/elisa-tech/wg-automotive/tree/master/Cluster_Display_Use_Case_v2) more "realistic",  but in the end the work was not completed while now the automotive WG focus is shifting.

Realistically, IMO, what we can do is to:

a) complete this RVM excercise e refine the Kernel analysis methodology with the OSEP wg

b) we could discuss internally in our WG what modification we could do (with the HW we have) and eventually propose to revisit the Telltale v1 completed TSC (however this should be double checked with Phillipp)

c) Nothing prevents us from analysing different aspects of Linux against different safety claims eirther in the LFSCS, OSEP WG and Arch WG. Respectively from
- technical and code enhancement perspective
- methodological perspective
- safety analysis perspective

Reply:
ok